sections/6-generic-programming.md

```diff
@@ -0,0 +1,63 @@
+## 6. Generic Programming
+
+<img src="https://raw.githubusercontent.com/monifu/scala-best-practices/master/assets/scala-logo-256.png"  align="right" width="128" height="128" />
+
+### 6.1. MUST NOT use automatic typeclass instance derivation
+
+When using [generic programming](https://github.com/milessabin/shapeless) or working with
+libraries based on it (such as [io.circe](https://circe.github.io/circe/codecs/auto-derivation.html)) avoid
+using **automatic derivation** features. An anti-pattern that often comes up look like this:
+
+```scala
+// Beware such imports!
+import io.circe.generic.auto._
+```
+
+or
+
+```scala
+import io.circe.generic.AutoDerivation
+
+object Model extends AutoDerivation {
+    case class GraphQLId(value: String)
+    case class User(name: String, login: String)
+    case class Comment(createdAt: OffsetDateTime, user: Option[User], commentText: Option[String])
+
+    // ... and many more classes
+}
+```
+
+In the example above `AutoDerivation` (`import io.circe.generic.auto._`) brings 
+into scope macros that generate at compile time `Encoder` and `Decoder` 
+instances case classes in scope:
+
+- It is hard to control what is being generated
+- It is hard to reason about what is being really used
+- Might take **a lot** of [compilation time](https://www.scala-lang.org/blog/2018/06/04/scalac-profiling.html) (in order to resolve an implicit a macro might get [expanded, resolved and thrown away](https://www.scala-lang.org/blog/2018/06/04/scalac-profiling.html#the-cost-of-implicit-macros))
```

Comment:
me being grammar nazi: `Might take **a long** [time to compile](...)`

Reply:
NP. More strictness is always appreciated. Will fix.